src/ast/ast_validations.c

```diff
@@ -474,15 +503,74 @@ static AST_Validation _Validate_MATCH_Clauses(const AST *ast, char **reason) {
 		AST_ReferredFunctions(match_clause, referred_funcs);
 	}
 
+	// Verify that no relation alias is also used to denote a node
+	for(uint i = start; i < end; i ++) {
+		const cypher_astnode_t *match_clause = cypher_ast_query_get_clause(ast->root, match_indices[i]);
+		res = _ValidateReusedIdentifiers(identifiers, cypher_ast_match_get_pattern(match_clause), reason);
+		if(res == AST_INVALID) goto cleanup;
+	}
+
+cleanup:
+	TrieMap_Free(identifiers, TrieMap_NOP_CB);
+
+	return res;
+}
+
+static AST_Validation _Validate_MATCH_Clauses(const AST *ast, char **reason) {
+	// Get the index of every MATCH clause within the AST.
+	uint *match_indices = AST_GetClauseIndices(ast, CYPHER_AST_MATCH);
+	uint match_count = array_len(match_indices);
+	// Return early if we don't have MATCH clauses.
+	if(match_count == 0) {
+		array_free(match_indices);
+		return AST_VALID;
+	}
+
+	AST_Validation res;
+	TrieMap *referred_funcs = NewTrieMap();
+
+	const cypher_astnode_t *return_clause = AST_GetClause(ast, CYPHER_AST_RETURN);
+	TrieMap *projections = _AST_GetReturnProjections(return_clause);
+
+	// Aliases are scoped by the WITH clauses within the query.
+	// If we have one or more WITH clauses, MATCH validations should be performed one scope at a time.
+	uint *query_scopes = AST_GetClauseIndices(ast, CYPHER_AST_WITH);
+	uint with_clause_count = array_len(query_scopes);
+	uint match_start = 0;
+	if(with_clause_count > 0) {
+		uint scope_start = 0;
+		for(uint i = 0; i < with_clause_count; i ++) {
+			uint match_end = match_start;
+			uint scope_end = query_scopes[i];
+			// Find the last MATCH clause within this scope.
+			while((match_end < match_count) && (match_indices[match_end] < scope_end)) match_end++;
+
+			// Perform validations on all MATCH clauses within this scope.
+			res = _ValidateMatchClausesInScope(ast, projections, referred_funcs, match_indices, match_start,
```

Comment:
Wouldn't you prefer to create an AST object for each scope and simply call _ValidateClauses on it?
If the cost of creating these sub ASTs isn't great I believe this will simplify the overall logic and make the validation bit more clear. thoughts?